spacy/training/example.pyx

```diff
@@ -150,27 +150,84 @@ cdef class Example:
                 self._y_sig = y_sig
                 return self._cached_alignment
 
-    def get_aligned(self, field, as_string=False):
-        """Return an aligned array for a token attribute."""
-        align = self.alignment.x2y
 
-        vocab = self.reference.vocab
-        gold_values = self.reference.to_array([field])
+    def _get_aligned_vectorized(self, align, gold_values):
+        # Fast path for Doc attributes/fields that are predominantly a single value,
+        # i.e., TAG, POS, MORPH.
+        x2y_single_toks = []
+        x2y_single_toks_i = []
+
+        x2y_multiple_toks = []
+        x2y_multiple_toks_i = []
+
+        # Gather indices of gold tokens aligned to the candidate tokens into two buckets.
+        #   Bucket 1: All tokens that have a one-to-one alignment.
+        #   Bucket 2: All tokens that have a one-to-many alignment.
+        for idx, token in enumerate(self.predicted):
+            aligned_gold_i = align[token.i]
+            if len(aligned_gold_i) == 1:
+                x2y_single_toks.append(aligned_gold_i.item())
+                x2y_single_toks_i.append(idx)
+            else:
+                x2y_multiple_toks.append(aligned_gold_i)
+                x2y_multiple_toks_i.append(idx)
+
+        # Map elements of the first bucket directly to the output array.
+        output = numpy.full(len(self.predicted), None)
+        output[x2y_single_toks_i] = gold_values[x2y_single_toks].squeeze()
+
+        # Collapse many-to-one alignments into one-to-one alignments if they
+        # share the same value. Map to None in all other cases.
+        for i in range(len(x2y_multiple_toks)):
+            aligned_gold_values = gold_values[x2y_multiple_toks[i]]
+
+            # If all aligned tokens have the same value, use it.
+            if len(set(list(aligned_gold_values))) == 1:
```

Comment:
I wonder if `groupby` isn't faster, since it doesn't involve intermediate set construction. See the `all_equal` recipe in

https://docs.python.org/3/library/itertools.html